src/memory_manager.rs

```diff
@@ -287,13 +396,108 @@ impl<M: Memory> MemoryManagerInner<M> {
+        }
+    }
+
+    fn load_layout_v2(memory: M, header: Header) -> Self {
+        let mut memory_size_in_buckets = vec![];
+        let mut number_of_used_buckets = 0;
+
+        for memory_size_in_pages in header.memory_sizes_in_pages.into_iter() {
+            let size_in_buckets = (memory_size_in_pages + header.bucket_size_in_pages as u64 - 1)
+                / header.bucket_size_in_pages as u64;
+            memory_size_in_buckets.push(size_in_buckets);
+            number_of_used_buckets += size_in_buckets;
+        }
+
+        const BYTE_SIZE_IN_BITS: usize = 8;
+        let buckets_index_size_in_bytes: usize =
+            (number_of_used_buckets as usize * BUCKET_ID_LEN_IN_BITS + (BYTE_SIZE_IN_BITS - 1))
+                / BYTE_SIZE_IN_BITS;
+
+        let mut buckets = vec![0; buckets_index_size_in_bytes];
+        memory.read(bucket_indexes_offset().get(), &mut buckets);
+
+        let buckets_decompressed = bytes_to_bucket_indexes(&buckets);
+
+        let mut memory_buckets = BTreeMap::new();
+
+        let mut last_occupied_bucket: i32 = -1;
+
+        let mut j = 0;
+
+        for (memory, size_in_buckets) in memory_size_in_buckets.into_iter().enumerate() {
+            let mut vec_buckets = vec![];
+            for _ in 0..size_in_buckets {
+                let bucket = buckets_decompressed[j];
+                last_occupied_bucket = std::cmp::max(bucket.0 as i32, last_occupied_bucket);
+                vec_buckets.push(bucket);
+                j += 1;
+            }
+            memory_buckets
+                .entry(MemoryId(memory as u8))
+                .or_insert(vec_buckets);
+        }
+
+        let mem_mgr = if last_occupied_bucket == -1 {
+            Self {
+                memory,
+                allocated_buckets: header.num_allocated_buckets,
+                bucket_size_in_pages: header.bucket_size_in_pages,
+                memory_sizes_in_pages: header.memory_sizes_in_pages,
+                memory_buckets,
+                freed_buckets: BTreeSet::new(),
+            }
+        } else {
+            let mut freed_buckets = BTreeSet::new();
+
+            for id in 0..last_occupied_bucket as u16 {
+                freed_buckets.insert(BucketId(id));
+            }
+
+            for id in buckets_decompressed.iter() {
+                freed_buckets.remove(id);
+            }
+
+            Self {
+                memory,
+                allocated_buckets: header.num_allocated_buckets,
+                bucket_size_in_pages: header.bucket_size_in_pages,
+                memory_sizes_in_pages: header.memory_sizes_in_pages,
+                memory_buckets,
+                freed_buckets,
+            }
+        };
+
+        // Update the header.
+        mem_mgr.save_header();
+
+        // Rewrite the stable store to be layout V2.
+        write(
+            &mem_mgr.memory,
+            bucket_indexes_offset().get(),
+            mem_mgr.get_bucket_ids_in_bytes().as_ref(),
+        );
```

Comment:
I will need to give it a look again, I do not remember.

Reply:
The initial idea was to avoid saving when freeing the memory, so here it was saved, but that approach was changed, and this left. I will remove this part.

Reply:
However, this saving can be good to add when loading layout V1. That way we will be sure that the next time it will be loaded as layout V2 even if no changes to memory are made. WDYT?